docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+    if err != nil {
+        return err
+    }
+
+    // check if user already has subscribe to this service
+    // if so renew the subscription
+    if subscription := GetUserSubscriptions(msg.Subscriber, msg.Name); subscription != nil {
+        if msg.Limit == -1 {
+            subscription.Limit = -1
+        } else {
+            subscription.Limit += msg.Limit
+        }
+        // set active to true
+        subscription.Active = true
+        StoreSubscription(subscriptionID, subscription)
+        return nil
+    }   
+    subscription := Subscription{
+        Name:       metadata.Name,
+        Subscriber: msg.Subscriber,
+        Limit:      msg.Limit,
+        LastPaid:   ctx.BlockTime,
+        Active:     true
+    }
+    StoreSubscription(subscriptionID, subscription)
+    // adds subscription to list of user subscriptions
+    AppendToUserSubscriptions(subscriptionID)
+    // push to back of DueQueue for this service
+    PushToBackOfDueQueue(msg.Name, subscriptionID)
+}
+```
+
+```go
+func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
+    subscription := GetUserSubscriptions(msg.Subscriber, msg.Name)
+    if subscription == nil {
+        return nil
+    }
+    subscription.Active = false
+    subscription.Limit = 0 // to ensure that renewal updates limit correctly
+    subscriptionID := hash(msg.Name|msg.Subscriber)
+    StoreSubscription(subscriptionID, subscription)
+}
+```
+
+```go
+func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+    metadata := GetMetaData(ctx, msg.Name)
+    if msg.Collector != metadata.Collector {
+        return InvalidCollectorErr
+    }
+    queue := GetDueQueue(msg.Name)
+    for i := 0; i < msg.Limit; i++ {
+        // top of the queue is "due"
+        subscription := GetSubscription(queue[0])
+        if subscription.LastPaid + metadata.Period >= ctx.BlockTime {
+            subscriptionID := PopOffFront(queue)
+            if subscription.Limit == 0 {
+                subscription.Active = false
+                StoreSubscription(subscriptionID, subscription)
+                continue
+            }
+            err := bank.Send(subscription.Subscriber, metadata.Collector, metadata.Amount)
+            if err != nil {
+                // insufficient funds. inactivate subscription
+                subscription.Active = false
+                subscription.Limit = 0
+            } else {
+                // update subscription after successful payment
+                subscription.LastPaid = ctx.BlockTime
+                if subscription.Limit != -1 {
+                    subscription.Limit--
+                }
+            }
+            // store updated subscription
+            StoreSubscription(subscriptionID, subscription)
+            // push paid subscription back to end of queue
+            PushToBack(queue, subscriptionID)
```

Comment:
this _only_ works if all the subscriptions have the same `Period`. Otherwise, you need to sort them again

Reply:
This is correct